### PR TITLE
Add register job and plan job features in backend

### PR DESCRIPTION
**Pull request by AlexandreAntoine**

### CODE REVIEW

**Comment by AlexandreAntoine** [comment]
#7 Issues: Create jobs through UI

**Comment by jippi** [comment]
Looks good (don't know Go though :))

Would it be too much to add `stop` as well? :)

**Comment by AlexandreAntoine** [comment]
I can add ```stop``` as well.

**Comment by rmessner** [comment]
I think that adding the `stop`now is too soon, because you can't start it again without an independant storage to keep the job definition.

Better making small steps, adding small features one by one.

**Comment by jippi** [comment]
@rmessner i would implement 'stop' as delete until the basic task management hits nomad (maybe in 0.6.x) - i've had more than once case where that has been desired :)

**Comment by jippi** [comment]
@AlexandreAntoine nice, do you plan to do the UI too?

**Comment by rmessner** [comment]
@jippi yes, we plan with @AlexandreAntoine to develop the UI for the Update/Manage job.

On the UI, the focus will be made to the driver config parameters which will be the common updates, and an advanced mode to edit all.

As soon as we have a first version, we will submit it.

**Comment by Nomon** [comment]
because this is not gated or optioned in any way it is forcing us to stop use the project. Letting this ui run or replace the jobs is unacceptable for our use :/

**Comment by jippi** [comment]
@Nomon gating it would make sense :) and afaik not terrible hard to do, have a READONLY env or something that hides all actions :)

File: frontend/src/containers/job.js [hunk 1]
```diff
@@ -36,6 +36,10 @@ class Job extends Component {
                     name: 'Raw',
                     path: 'raw',
                 },
+                {
```
Comment (jippi) [inline frontend/src/containers/job.js]: can you put this before `raw`? :)

File: frontend/src/router.js [hunk 2]
```diff
@@ -59,6 +60,7 @@ const AppRouter = ({ history }) =>
         <Route path="/jobs/:jobId/tasks" component={ JobTasks } />
         <Route path="/jobs/:jobId/taskGroups" component={ JobTaskGroups } />
         <Route path="/jobs/:jobId/raw" component={ JobRaw } />
+        <Route path="/jobs/:jobId/manage" component={ JobManage } />
```
Comment (jippi) [inline frontend/src/router.js]: same here :)

**Comment by AlexandreAntoine** [comment]
We can add the ```stop``` which set the taskGroup's ```count``` to 0. This way the job isn't destroy.

**Comment by jippi** [comment]
i like it, maybe call it `pause` or something, `stop` should be long-term "nomad stop" compatible :)

File: backend/main.go [hunk 3]
```diff
@@ -52,6 +55,8 @@ var (
 		"Overrides the NOMAD_ADDR environment variable if set. "+flagDefault(defaultConfig.Address))
 	flagListenAddress = flag.String("web.listen-address", "",
 		"The address on which to expose the web interface. "+flagDefault(defaultConfig.ListenAddress))
+	flagReadOnly = flag.Bool("readonly", true,
+		"The possibility to execut or not writing optrations. "+flagDefault(strconv.FormatBool(defaultConfig.ReadOnly)))
```
Comment (jippi) [inline backend/main.go]: `execute`

File: backend/connection.go [hunk 4]
```diff
@@ -140,6 +148,69 @@ func (c *Connection) Handle() {
...
+	}
+	jobjson := action.Payload.(string)
+	runjob := api.Job{}
+	json.Unmarshal([]byte(jobjson), &runjob)
+
+	logger.Infof("Started run job with id: %s", runjob.ID)
+
+	job, _, err := c.hub.nomad.Client.Jobs().Register(&runjob, nil)
+	if err != nil {
+		logger.Errorf("connection: unable to register job : %s", err)
```
Comment (jippi) [inline backend/connection.go]: shouldn't we bubble this up to the end-user?

File: backend/connection.go [hunk 5]
```diff
@@ -140,6 +148,69 @@ func (c *Connection) Handle() {
...
+	}
+	jobjson := action.Payload.(string)
+	stopjob := api.Job{}
+	json.Unmarshal([]byte(jobjson), &stopjob)
+
+	logger.Infof("Started stop job with id: %s", stopjob.ID)
+
+	job, _, err := c.hub.nomad.Client.Jobs().Deregister(stopjob.ID, nil)
+	if err != nil {
+		logger.Errorf("connection: unable to stop job : %s", err)
```
Comment (jippi) [inline backend/connection.go]: same here

File: backend/connection.go [hunk 6]
```diff
@@ -140,6 +148,69 @@ func (c *Connection) Handle() {
...
+	}
+	jobjson := action.Payload.(string)
+	planjob := api.Job{}
+	json.Unmarshal([]byte(jobjson), &planjob)
+
+	logger.Infof("Started plan job with id: %s", planjob.ID)
+
+	plan, _, err := c.hub.nomad.Client.Jobs().Plan(&planjob, true, nil)
+	if err != nil {
+		logger.Errorf("connection: unable to plan job : %s", err)
```
Comment (jippi) [inline backend/connection.go]: and here

**Comment by jippi** [comment]
@AlexandreAntoine what is the status on this PR ? :) 

Do you have more work planned for it?

**Comment by AlexandreAntoine** [comment]
I am going to improve what I have done and add an advanced mode which directly edit the json configuration.

**Comment by jippi** [comment]
@AlexandreAntoine okay, I'm sorry up-front, but probably most of the react stuff will not merge into master cleanly once  #153  is merged in. I've spend a week time completely refactored it. 

I dont know will be a huge pain in the ass for you and this PR, so if you need any help porting it over, please let me know. I have not touched the Go code, so that should merge in just fine. 

The stuff you've done in React seem to be mostly isolated, so it should be mostly just putting the file in the right place and add the route back again.

if you need any help, please let me know !

**Comment by jippi** [comment]
@AlexandreAntoine did you see the comment https://github.com/iverberk/nomad-ui/pull/128#issuecomment-265420108 re the new UI? :)

**Comment by AlexandreAntoine** [comment]
@jippi I'm going to do the merge. The code I have written is mostly isolated, so I think it will be easy to do the merge. May be I will create a Knew branch one #153 is merged.

**Comment by jippi** [comment]
@AlexandreAntoine I've merged in my mege-change to master now - seem to be a lot of merge conflicts, I'm so sorry for the pain it may cause your productivity while continuing your work on the PR

**Comment by ndriessen** [comment]
Any idea when this can be merged in and released? I urgently need this and since it seems to be implemented I don' want to build something myself...

**Comment by AlexandreAntoine** [comment]
I have started to merge by hand but everything has changed in backend during the merge, I have also encountered problems with the "react-s-alert" plugin and other stuff with css. So I have decided to stay on the old version for now.

**Comment by jippi** [comment]
@AlexandreAntoine I think the usage for `react-s-alert`  should be replaced with http://www.material-ui.com/#/components/snackbar - which is native to the already existing plugins used in the rewrite

Maybe http://www.material-ui.com/#/components/dialog too if you need to confirm anything

**Comment by jippi** [comment]
@AlexandreAntoine can you share your partial merge in a new PR, maybe me or @ndriessen can wrap it up :)

**Comment by ndriessen** [comment]
Hey @jippi I would love to help out, but first would be learning both go and react :) 
Is there a quick guide on how to get started to setup/build this project locally (using a macbook) ?

**Comment by jippi** [comment]
@ndriessen I've tried to outline the process I use here: https://github.com/jippi/hashi-ui/blob/master/DEVELOPMENT.md 

I'm still new to Go as well (<2months :P) 

Otherwise swing by the hashi-ui gitter channel and I can try to help out if things go sideways

**Comment by jippi** [comment]
@AlexandreAntoine is this the primary UI change you've done ?

image


any other visual changes?

**Comment by ndriessen** [comment]
@jippi thanks I missed that file on the dev setup... I have a few other priorities to finish this week, but I'm gonna look into getting the whole dev setup working next week... I'll give you a shout if I run into anything blocking

**Comment by jippi** [comment]
I've updated master to expose `NOMAD_READ_ONLY`  in JS, so can be checked using `window.NOMAD_READ_ONLY` inside React :) 

Thats at least one feature / change less needed to bring this PR into the core.

**Comment by AlexandreAntoine** [comment]
I've a lot of work to do. Maybe I'm going to merge it tomorrow.

**Comment by jippi** [comment]
I've finished up part of the job modifications now - simple controls for scale up/down/stop a task group

image

**Comment by jippi** [comment]
I've added support for editing the full job JSON like this

image

Commit: https://github.com/jippi/hashi-ui/commit/852c5db9c8c956f4dae0c17ff96a1d2f9433691a

I'm closing this issue as they seem to be on feature parity now (?)

Thanks 1000x for your contribution @AlexandreAntoine - it was a great guide for how to implement this feature! I'm so sorry my major refactoring got in the way of merging your contribution :(

**Comment by AlexandreAntoine** [comment]
Thanks for merging these features. You did it a better way than if I had done it.

**Pull request closed without merging**